app/user.go

```diff
@@ -0,0 +1,597 @@
+	if err != nil {
+		return nil, err
+	}
+	return &UserClient{
+		client: authservice.NewAuthServiceClient(conn),
+		ctx:    ct,
+	}, nil
+}
+
+func (c *UserClient) listUsers(
+	namespace string,
+) error {
+	totalRes := &authservice.GetUsersResponse{}
+	pageToken := ""
+	for {
+		res, err := c.client.GetUsers(c.ctx, &authservice.GetUsersRequest{
+			PageToken: pageToken,
+			Namespace: namespace,
+		})
+		if err != nil {
+			return err
+		}
+		totalRes.Users = append(totalRes.Users, res.Users...)
+		// Check if we should continue paging
+		pageToken = res.NextPageToken
+		if len(pageToken) == 0 {
+			return PrintProto(totalRes)
+		}
+	}
+}
+
+func (c *UserClient) getUser(userID, userEmail string) (*auth.User, error) {
+	if userID == "" && userEmail == "" {
+		return nil, fmt.Errorf("both user-id and user-email not set")
+	}
+
+	if userID != "" && userEmail != "" {
+		return nil, fmt.Errorf("both user-id and user-email set")
+	}
+	res, err := c.client.GetUser(c.ctx, &authservice.GetUserRequest{
+		UserId:    userID,
+		UserEmail: userEmail,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if res.User == nil || res.User.Id == "" {
+		// this should never happen, the server should return an error when the user is not found
+		return nil, fmt.Errorf("invalid user returned by server")
+	}
+	return res.User, nil
+}
+
+func (c *UserClient) getUserAndRoles(userID, userEmail string) (*auth.User, []*auth.Role, error) {
+	user, err := c.getUser(userID, userEmail)
+	if err != nil {
+		return nil, nil, err
+	}
+	var pageToken string
+	var roles []*auth.Role
+	for {
+		res, err := c.client.GetRoles(c.ctx, &authservice.GetRolesRequest{
+			PageToken: pageToken,
+			UserId:    user.Id,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+		roles = append(roles, res.Roles...)
+		// Check if we should continue paging
+		pageToken = res.NextPageToken
+		if len(pageToken) == 0 {
+			break
+		}
+	}
+
+	return user, roles, nil
+}
+
+func (c *UserClient) inviteUsers(
+	ctx *cli.Context,
+	emails []string,
+	namespacePermissions []string,
+	accountRole string,
+) error {
+	if len(accountRole) == 0 {
+		return errors.New("account role required for inviting new users")
+	}
+
+	// the role ids to invite the users for
+	var roleIDs []string
+
+	// first get the required account role
+	role, err := getAccountRole(c.ctx, c.client, accountRole)
+	if err != nil {
+		return err
+	}
+	roleIDs = append(roleIDs, role.GetId())
+
+	// get any optional namespace permissions
+	if len(namespacePermissions) > 0 {
+		npm, err := toUserNamespacePermissionsMap(namespacePermissions)
+		if err != nil {
+			return err
+		}
+		nsRoles, err := getNamespaceRolesFromMap(c.ctx, c.client, npm)
+		if err != nil {
+			return err
+		}
+		for _, nsRole := range nsRoles {
+			roleIDs = append(roleIDs, nsRole.GetId())
+		}
+	}
+
+	// Invite users to given roles
+	req := &authservice.InviteUsersRequest{
+		Specs:     make([]*auth.UserSpec, len(emails)),
+		RequestId: ctx.String(RequestIDFlagName),
```

Comment:
Should `RequestIDFlagName` be required, since we directly use it in most of these calls?

Reply:
the server assigns one automatically if its not supplied by creating a new UUID. What we could do if its not specified is generate the UUID ourselves maybe?

Reply:
Server assigning one sounds best, that makes sense!